### ISSUE TITLE
Incremental text sync issues

### ISSUE THREAD

**Issue by artempyanykh** [emacs-lsp/lsp-mode#4013]
### Thank you for the bug report

- [X] I am using the latest version of `lsp-mode` related packages.
- [X] I checked [FAQ](https://emacs-lsp.github.io/lsp-mode/page/faq/) and [Troubleshooting](https://emacs-lsp.github.io/lsp-mode/page/troubleshooting/) sections
- [ ] You may also try reproduce the issue using clean environment using the following command: `M-x lsp-start-plain`

### Bug description

When incremental text sync is on, lsp-mode sends changes that violate the current state of the document. With the repro I described below, lsp-mode send the following notifications to the server:
```
[Trace - 01:31:08 pm] Sending notification 'textDocument/didOpen'.
Params: {
  "textDocument": {
    "uri": "file:///Users/arr/personal-notes/test.md",
    "languageId": "markdown",
    "version": 182,
    "text": "# Test\n\nLooks\n* LOOK_INTO\n* LOOK_\n\n## Baz\n\nBar Baz\n"
  }
}


[Trace - 01:31:24 pm] Sending notification 'textDocument/didChange'.
Params: {
  "textDocument": {
    "uri": "file:///Users/arr/personal-notes/test.md",
    "version": 183
  },
  "contentChanges": [
    {
      "range": {
        "start": {
          "line": 4,
          "character": 2
        },
        "end": {
          "line": 4,
          "character": 11
        }
      },
      "rangeLength": 9,
      "text": "LOOK_INTO"
    }
  ]
}


[Trace - 01:31:24 pm] Sending notification 'textDocument/didChange'.
Params: {
  "textDocument": {
    "uri": "file:///Users/arr/personal-notes/test.md",
    "version": 184
  },
  "contentChanges": [
    {
      "text": "# Test\n\nLooks\n* LOOK_INTO\n* LOOK_INTO\n\n## Baz\n\nBar Baz\n"
    }
  ]
}
```

Note that the 2nd request says that it wants to change the range `(4,2) - (4,11)` but column 11 is outside of the documents range. What's weird is that lsp-mode sends another didChange notification right after that with the full document's text.

### Steps to reproduce

1] Open a markdown document with the following content with [Marksman](https://github.com/artempyanykh/marksman) setup:
```
# Test

Looks
* LOOK_INTO
* LOOK_

## Baz

Bar Baz
```

2] Position your cursor on line 5 right after `LOOK_`
3] Trigger dabbrev-expand

### Expected behavior

Incremental text sync shouldn't send borked change notifications because these are unrecoverable by the server.

### Which Language Server did you use?

marksman

### OS

MacOS

### Anything else?

I'm using lsp-mode 20230401.434, Emacs 28.2, darwin

**Comment by yyoncho** [comment]
Indeed. As a workaround I suggest using M-x hippie-expand (which will try dabbrev but in a different way that avoids this issue).

**Comment by artempyanykh** [comment]
Thanks for the workaround @yyoncho!

I'm actually planning to disable incremental text sync in Marksman for the time being, because I see problems reported from users of other editors too and the LSP protocol doesn't help much in terms of debugging/triaging this types of issues. Perhaps you might be interested to chime in on this meta issue https://github.com/microsoft/language-server-protocol/issues/1706.

**Comment by yyoncho** [comment]
that particular issue is between emacs and lsp-mode (the code is very tricky and patched a lot)

**Comment by artempyanykh** [comment]
@yyoncho thank you, I understand. My point was that as a maintainer of an LSP server (Marksman in this case) I get bug reports from users of different editors/lsp-clients. Triaging and debugging such issues is time-consuming especially for editors that I don't use personally. https://github.com/microsoft/language-server-protocol/issues/1706 is about adding more data to the LSP protocol to allow stronger validation on the server side and consequently simplify debugging.
